### ISSUE TITLE
[Feature] Full system encryption using GELI

### ISSUE THREAD

**Issue by no-usernames-left** [ghostbsd/issues#68]
FreeBSD (including version 13.0-RELEASE) supports full system encryption out of the box using GELI.

GhostBSD should incorporate this functionality, as it's based on FreeBSD; indeed, GhostBSD-21.04.27 is based on FreeBSD 13.0-RELEASE. (It looks like most of the work needed might be in the installer.)

**Comment by ericbsd** [comment]
issue moved from ghostbsd/ghostbsd-src to ghostbsd/issues

**Comment by vermaden** [comment]
> > There is easy workaround:
> 
> Maybe if you know ZFS well, I'm not sure I know what you mean. 🤔

Which part? I can explain.

**Comment by anonymous-cupholder** [comment]
Let's look more deeply about which encryption method meets end user needs best - GELI encryption or ZFS encryption.

Keep in mind when implementing GELI is it is an 'either or' option. In other words, select GELI encryption or ZFS encryption, but not both. When selecting GELI we need to grey-out the ZFS encryption option when implemented. Also, whether selecting GELI encryption or ZFS encryption, the selection will affect a backup and restore solution.  

**BACKGROUND**

GELI Encryption:
        Strengths: Full disk encryption, compatibility with all filesystems, separate key management.
        Weaknesses: Performance overhead, management complexity, limited ZFS integration, challenging boot environment setup, complex key management.

ZFS Native Encryption:
        Strengths: Granular dataset-level encryption, integrated with ZFS features, potentially better performance, straightforward backup with ZFS send/receive.
        Weaknesses: Granularity adds complexity, limited to ZFS filesystems, performance impact, complex key management, requires compatible backup solutions.

**IMPACT TO BACKUP SOLUTIONS**

Here's a summary of the impact of selecting either GELI or ZFS encryption on data backup and restoration processes:

GELI Encryption:

Backup:

    Decrypted State:
        Backup software can access files as if they are unencrypted when the filesystem is mounted and decrypted.
        Standard file-level backup tools can be used without modification.
    Encrypted State:
        Backup the entire encrypted volume as raw data.
        This approach is less common for regular backups and more suited for disaster recovery scenarios.

Restore:

    Decrypted State:
        Restore files using standard file-level restoration tools.
        Re-encryption occurs automatically as data is written back to the encrypted volume.
    Encrypted State:
        Restore the entire encrypted volume.
        Requires restoring the raw encrypted data and then decrypting it to access the files.

Considerations:

    Performance: Decrypting data for backups may introduce performance overhead.
    Complexity: Managing full disk encryption, including boot environments, can be complex.
    Compatibility: Most backup software works without issue when the filesystem is mounted and decrypted.

ZFS Native Encryption:

Backup:

    Decrypted State:
        Backup software can access files as usual when datasets are mounted and decrypted.
        Standard file-level backup tools can be used.
    Encrypted State:
        ZFS supports sending encrypted datasets using zfs send.
        Encrypted backups can be transferred and stored securely without needing to decrypt them first.

Restore:

    Decrypted State:
        Restore files using standard file-level restoration tools.
        Re-encryption is handled by ZFS as data is written back to the encrypted datasets.
    Encrypted State:
        Restore encrypted datasets using zfs receive.
        Maintains encryption properties and does not require decryption during transfer.

Considerations:

    Performance: Optimized for ZFS, but encryption/decryption still adds some overhead.
    Integration: Seamlessly integrates with ZFS features like snapshots, replication, and dataset management.
    Granularity: Offers granular encryption at the dataset level, allowing selective encryption of data.

Summary of Impact:
GELI Encryption:

    Backup: Can be done in either decrypted or encrypted state. Decrypted state is more straightforward for regular backups.
    Restore: Requires re-encryption if restored from decrypted backups. Raw encrypted backups need the entire volume restored and decrypted.
    Complexity: Higher due to full disk encryption, especially for boot environments and mixed filesystems.

ZFS Native Encryption:

    Backup: Supports both decrypted and encrypted states. Encrypted backups using zfs send are efficient and secure.
    Restore: Encrypted datasets can be restored while maintaining encryption properties. Simplifies the process with ZFS tools.
    Integration: Better integration with ZFS features, offering more flexibility and ease of use.

In conclusion, if you primarily use ZFS and want to leverage its features, ZFS native encryption is likely more efficient and easier to manage for both backups and restores. If full disk encryption is required or you use a mix of filesystems, GELI may be necessary, but it adds complexity and potential performance overhead.

**Comment by ericbsd** [comment]
If I remember correctly, FreeBSD does not support to boot zfs encrypted root.

**Comment by vermaden** [comment]
> If I remember correctly, FreeBSD does not support to boot zfs encrypted root.

Exactly.

That is why - today - its GELI or nothing.

... or just encrypt `/home/USER` dir with ZFS - its even supported in `adduser(8)` script now.

**Comment by vermaden** [comment]
> Let's look more deeply (...)

Thank You for summarizing ups and downs between GELI and ZFS encryption.

**Comment by anonymous-cupholder** [comment]
This is what I've learned. The best workaround for booting FreeBSD with an encrypted ZFS root filesystem involves using an unencrypted `/boot` partition to store the necessary boot files. This approach ensures that the system can load the bootloader and kernel before unlocking the encrypted ZFS root. Here’s a step-by-step outline of this method:

Create Partitions:
   - Create a small, unencrypted UFS or ZFS partition for `/boot`.
   - Create a larger ZFS partition for the root filesystem with encryption enabled.

   gpart create -s gpt ada0
   gpart add -t efi -a 1m -s 256m -l efi ada0
   gpart add -t freebsd-boot -a 1m -s 512k -l boot ada0
   gpart add -t freebsd-ufs -a 1m -s 2G -l bootfs ada0
   gpart add -t freebsd-zfs -a 1m -l zroot ada0

2. Format the Partitions:
   - Format the EFI and UFS partitions.
   - Create the ZFS pool with encryption.

   newfs_msdos /dev/gpt/efi
   newfs /dev/gpt/bootfs
   zpool create -o ashift=12 -O encryption=on -O keyformat=passphrase zroot /dev/gpt/zroot

3. Install the Base System:
   - Mount the filesystems and install the base system and kernel to the ZFS pool.
   - Copy the necessary bootloader files to the UFS partition.

   zfs create zroot/ROOT
   mount -t ufs /dev/gpt/bootfs /mnt
   tar -xpf /path/to/base.txz -C /mnt
   tar -xpf /path/to/kernel.txz -C /mnt
   echo 'zfs_load="YES"' >> /mnt/boot/loader.conf

4. Configure the Bootloader:
   - Copy the bootloader to the EFI partition.
   - Configure the loader to point to the root ZFS pool.

   mkdir -p /mnt/boot/efi/boot
   cp /mnt/boot/loader.efi /mnt/boot/efi/boot/bootx64.efi

5. Finalize the Configuration:
   - Set the ZFS boot filesystem.
   - Unmount and reboot the system.

   zpool set bootfs=zroot/ROOT zroot
   umount /mnt
   reboot

This workaround ensures that the system can boot from the unencrypted `/boot` partition and then unlock and mount the encrypted root filesystem.

For detailed instructions and community discussions, you can refer to the following sources:

1. [FreeBSD Forums - Quest for Unencrypted `/boot` and ZFS Native Encrypted Root](https://forums.freebsd.org/threads/solved-the-quest-for-unencrypted-boot-and-zfs-native-encrypted-root.81457/)【7†source】.
2. [Reddit Discussion on Booting FreeBSD with Encrypted ZFS Root](https://www.reddit.com/r/freebsd/comments/pvkm0g/booting_to_freebsd_root_on_encrypted_zfs_dataset/)【9†source】.
3. [FreeBSD Forums - Loader.efi and Encrypted ZFS Root](https://forums.freebsd.org/threads/solved-loader-efi-cannot-boot-native-encrypted-zfs-root.79958/)【10†source】.

**Comment by vermaden** [comment]
... and this is the first step to loose one of the best FreeBSD features **- ZFS Boot Environments**.

**Comment by anonymous-cupholder** [comment]
Here is a suggestion for implementing GELI.

### 1. `ghostbsd-build`

**No changes needed to `build.sh`.**

### 2. Adding GELI to Package Lists

**File: packages/mate**

```
geli
# Other existing packages...
```

**File: packages/xfce**

```
geli
# Other existing packages...
```

**File: packages/mate_oem**

```
geli
# Other existing packages...
```

### 3. `gbi` (GhostBSD Installer)

**File: src/lib/installer.py**

```
import subprocess

# Existing imports and code...

def setup_geli_encryption(partition, passphrase):
    try:
        # Initialize GELI on the specified partition
        subprocess.run(['geli', 'init', '-s', '4096', partition], check=True)
        # Attach the GELI partition with the passphrase
        process = subprocess.Popen(['geli', 'attach', '-j', '-', partition], stdin=subprocess.PIPE)
        process.communicate(input=passphrase.encode())
        process.wait()
        # Format the new GELI device
        new_partition = f"{partition}.eli"
        subprocess.run(['newfs', new_partition], check=True)
    except subprocess.CalledProcessError as e:
        print(f"Error setting up GELI encryption: {e}")
        raise  # Re-raise the exception after logging it for further handling
    except Exception as e:
        print(f"Unexpected error: {e}")
        raise

# Existing installer logic...

# Add GELI encryption step if selected
if user_selected_geli_encryption:
    setup_geli_encryption(selected_partition, user_passphrase)
```

**File: src/ui/installer_ui.py**

```
import gi
gi.require_version('Gtk', '3.0')
from gi.repository import Gtk

# Existing imports and code...

class InstallerUI:
    def __init__(self):
        self.builder = Gtk.Builder()
        self.builder.add_from_file("installer.ui")
        
        # Existing UI elements...
        
        self.geli_encryption_checkbox = Gtk.CheckButton(label="Enable GELI Encryption")
        self.geli_encryption_checkbox.connect("toggled", self.on_geli_encryption_toggled)
        self.grid.attach(self.geli_encryption_checkbox, 0, 5, 1, 1)

        self.geli_passphrase_entry = Gtk.Entry()
        self.geli_passphrase_entry.set_visibility(False)
        self.grid.attach(self.geli_passphrase_entry, 1, 5, 1, 1)

    def on_geli_encryption_toggled(self, widget):
        if widget.get_active():
            self.geli_passphrase_entry.set_sensitive(True)
        else:
            self.geli_passphrase_entry.set_sensitive(False)

    # Existing methods...

# Main function to start the installer
def main():
    app = InstallerUI()
    Gtk.main()

if __name__ == "__main__":
    main()
```

### 4. `pc-sysinstall`

**File: pc-sysinstall/backend/functions-disk.sh**

```
#!/bin/sh
# pc-sysinstall disk functions

# Existing functions...

setup_geli() {
    local partition=$1
    local passphrase=$2

    # Initialize GELI
    echo "$passphrase" | geli init -s 4096 /dev/"$partition" -J -
    # Attach GELI partition
    echo "$passphrase" | geli attach /dev/"$partition" -j -

    # Format the GELI encrypted partition
    newfs /dev/"$partition".eli
}

# Example usage
if [ "$GELI_ENCRYPTION" = "YES" ]; then
    setup_geli "$PARTITION" "$GELI_PASSPHRASE"
fi

# Additional functions...
```

**File: pc-sysinstall/conf/pc-sysinstall.conf**

```
# pc-sysinstall configuration

# Existing configuration options...

GELI_ENCRYPTION="YES"
GELI_PASSPHRASE="user_provided_passphrase"
```

**Comment by anonymous-cupholder** [comment]
Since ZFS encryption is still an immature solution. Implementing GELI encryption would be our only solution.

If GELI encryption creates breakage to other applications (e.g. backup station), then that would be an issue to consider.

**Comment by ericbsd** [comment]
GELY is already implemented but doesn't work properly.

**Comment by ghostfixer** [comment]
FreeBSD’s bootloader cannot read encrypted datasets. This means the boot environment (/boot) must remain unencrypted. However, you can encrypt /, /home, etc., but /boot must be in a separate unencrypted dataset or partition. This is trivial to implement.

**Comment by vermaden** [comment]
> GELY is already implemented but doesn't work properly.

Which part does not work properly?

**Comment by ericbsd** [comment]
> > GELY is already implemented but doesn't work properly.
> 
> Which part does not work properly?

It does not boot.

**Comment by vermaden** [comment]
> GhostBSD uses ZFS Boot Environments but it does not allow 'direct' install on GELI encryption.
> 
> There is easy workaround:
> 
>     * Install FreeBSD 14.0-STABLE with ZFS on GELI.
> 
>     * Install GhostBSD 24.01.1 (Based on FreeBSD 14.0-STABLE) in VM (Bhyve/Virtualbox/...).
> 
>     * ZFS send GhostBSD BE to FreeBSD machine.
> 
>     * Activate copied GhostBSD BE.
> 
>     * Reboot in that GhostBSD BE.
> 
> 
> Hope that helps.

Everytime I did this - it just worked.

**Comment by grahamperrin** [comment]
I hid https://github.com/ghostbsd/issues/issues/68#issuecomment-1806915737 as off-topic. Better discuss workarounds, and alternatives to GELI, in the forums: 

* https://forums.ghostbsd.org/

– [Full Disk Encryption](https://forums.ghostbsd.org/d/188-full-disk-encryption) might be ideal. It's reasonable to treat root-on-ZFS (OpenZFS) – part of a disk – as _full_ disk, loosely speaking, for this type of discussion.

### GELI

https://github.com/ghostbsd/pc-sysinstall

<https://github.com/search?q=repo%3Aghostbsd%2Fpc-sysinstall%20GELI&type=code>

Beyond boot: 

* encrypted swap should be `late` in fstab
* rc.conf `dumpdev` should specify the correct device.

dumpon(8) <https://man.freebsd.org/cgi/man.cgi?query=dumpon&sektion=8&manpath=freebsd-current>

fstab(5) <https://man.freebsd.org/cgi/man.cgi?query=fstab&sektion=5&manpath=freebsd-current>

rc.conf(5) <https://man.freebsd.org/cgi/man.cgi?query=rc.conf&sektion=5&manpath=freebsd-current>

----

In my (FreeBSD) case: 

```text
grahamperrin@mowa219-gjp4-zbook-freebsd ~> sysrc dumpdev
dumpdev: /dev/ada1p2
grahamperrin@mowa219-gjp4-zbook-freebsd ~> lsblk /dev/ada1
DEVICE         MAJ:MIN SIZE TYPE                                    LABEL MOUNT
ada1             0:130 932G GPT                                         - -
  ada1p1         0:137 260M efi                                   gpt/efi /boot/efi
  <FREE>         -:-   1.0M -                                           - -
  ada1p2         0:139  16G freebsd-swap                 gpt/freebsd-swap SWAP
  ada1p2.eli     1:24   16G freebsd-swap                                - SWAP
  ada1p3         0:141 915G freebsd-zfs                   gpt/freebsd-zfs <ZFS>
  ada1p3.eli     0:147 915G -                                           - -
  <FREE>         -:-   708K -                                           - -
grahamperrin@mowa219-gjp4-zbook-freebsd ~> 
```

**Comment by lawmanuk** [comment]
Just replying to add I'm also seeking full disk encryption as a huge priority for me, which I had on linux but don't have on gbsd which I just installed.

> Because it is not a GUI installer.

Would it be easier if GBSD had safety option to allow a text installer, to utilise some features from FBSD installer?

It would also help to bypass the problems of devices like mine which can't boot into video mode on the installer.    I had to press ctrl-c after loading system into memdisk then sudo xconfig setup to switch to vesa.

**Comment by ryannoblett** [comment]
I also want to add a vote for full disk encryption at installation. Not having it is a deal-breaker. I'd prefer ZFS encryption, but GELI would also be fine, either way it's absolutely required. 

From the technical side, a separate, unencrypted boot partition that then loads the kernel and mounts the encrypted ZFS root, is the usual way to do ZFS encryption. I couldn't get it to work with GhostBSD post-install despite significant effort, and the installer doesn't offer the required amount of partitioning flexibility to configure manually at install time either.

**Comment by lawmanuk** [comment]
Just wanted to add some detail to the request please.   Below would be extremely helpful.

1) Encryption target options:


a) Encrypt root partition only (with efi and no unencrypted boot partition) - so we can dual boot which is what I REALLY need while getting comfortable with BSD over linux.

b) Full disk encryption for those that need it - I imagine most people will need dual boot



2) Encryption method option:


a) Geli encryption - I prefer this so it hides metadata

b) ZFS encryption - Ryan requested this so adding but it leaks metadata so not preferred at all.

c) Luks for data partitions as an option when dual booting linux - I know this might be impossible but just adding as a wishlist to make life easier as I have a data partition on luks, so could access same data from both linux/zfs if dual booting.   I know linux also has zfs option but less easy to use/setup and not comfortable with that yet.

d) Shufflecake encryption - This is the ideal utopia as the nextgen veracrypt for linux, but I imagine its a long way off for BSD.  Would be heaven if ever added to BSD.
https://shufflecake.net/

Thanks for considering

**Comment by ericbsd** [comment]
<img width="806" height="534" alt="Image" src="https://github.com/user-attachments/assets/5aab9d86-31c8-4270-854e-3a6b342b0940" />

**Comment by lawmanuk** [comment]
Can this implementation only geli encrypt the partition reserved for ghostbsd (rather than entire disk)?

This way it will allow dualboot to my fully encrypted Linux partition as I transition slowly over to ghostbsd.

Only the EFI partition should be unencrypted. 

Thanks

**Comment by vermaden** [comment]
> Only the EFI partition should be unencrypted.


That is how GELI works - you encrypt just the partition 'under' ZFS pool.

Example:

```
% lsblk nda0   
DEVICE         MAJ:MIN  SIZE TYPE                                    LABEL MOUNT
nda0             0:95   477G GPT                                         - -
  nda0p1         0:97   477G efi                              gpt/efiboot0 -
  nda0p2         0:99   477G freebsd-boot                     gpt/gptboot0 -
  <FREE>         -:-    492K -                                           - -
  nda0p3         0:101  477G freebsd-swap                        gpt/swap0 -
  nda0p4         0:103  477G freebsd-zfs                   gpt/freebsd-zfs <ZFS>
  nda0p4.eli     0:110  477G zfs                                         - -
  <FREE>         -:-    324K -    
```

**Comment by ericbsd** [comment]
> Can this implementation only geli encrypt the partition reserved for ghostbsd (rather than entire disk)?
> 
> This way it will allow dualboot to my fully encrypted Linux partition as I transition slowly over to ghostbsd.

@lawmanuk, yes, it can be done, but that would need to be a new ticket. Since it will big undertaking to integrate that in the partition editor. The ZFS Configuration had the code for it already, it was just hidden.
